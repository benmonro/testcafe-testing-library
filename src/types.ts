import { Config, BoundFunction, queries } from "@testing-library/dom";

export type Options = Pick<Config, "testIdAttribute">;

export type TestcafeBoundFunction<T> = (
  ...params: Parameters<BoundFunction<T>>
) => SelectorPromise;

export type TestcafeBoundFunctions<T> = {
  [P in keyof T]: TestcafeBoundFunction<T[P]>;
};
