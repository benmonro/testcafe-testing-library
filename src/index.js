/* eslint-disable no-eval */
/* eslint-disable no-new-func */
import fs from 'fs'
import path from 'path'
import {ClientFunction, Selector} from 'testcafe'
import {queries} from 'dom-testing-library'

const LIBRARY_UMD_PATH = path.join(
  './node_modules',
  'dom-testing-library/dist/dom-testing-library.umd.js',
)
const LIBRARY_UMD_CONTENT = fs.readFileSync(LIBRARY_UMD_PATH).toString()

export const addTestcafeTestingLibrary = async t => {
  // eslint-disable-next-line
  const inject = ClientFunction(
    () => {
      // eslint-disable-next-line no-undef
      window.eval(script)
      window.TestCafeTestingLibrary = {}
    },
    {
      dependencies: {script: LIBRARY_UMD_CONTENT},
    },
  )

  await inject.with({boundTestRun: t})()
}

Object.keys(queries).forEach(queryName => {
  module.exports[queryName] = Selector(
    new Function(
      `
      return DomTestingLibrary.${queryName}(document.body, ...arguments);
      `,
    ),
  )
})

export const within = async sel => {
  await ClientFunction(
    new Function(
      ` 
    const elem = document.querySelector("${sel}");
    window.TestCafeTestingLibrary["within_${sel}"] = DomTestingLibrary.within(elem);

    `,
    ),
  )()
  const container = {}

  Object.keys(queries).forEach(queryName => {
    container[queryName] = Selector(
      new Function(
        `return window.TestCafeTestingLibrary["within_${sel}"].${queryName}(...arguments)`,
      ),
    )
  })

  return container
  // const container = {}

  // Object.keys(queries).forEach(queryName => {
  //   container[queryName] = Selector(
  //     new Function(
  //       `return DomTestingLibrary.within(document.querySelector("${sel}")).${queryName}(...arguments)`,
  //     ),
  //   )
  // })

  // return container
}
