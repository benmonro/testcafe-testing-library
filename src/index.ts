import { ClientFunction, Selector } from "testcafe";
import { queries } from "@testing-library/dom";
import type {
  Options,
  TestcafeBoundFunction,
  TestcafeBoundFunctions,
} from "./types";

declare global {
  interface Window {
    TestingLibraryDom: typeof queries;
  }
}

const SELECTOR_TYPE = Selector("")().constructor.name;

const withinSelectors = Object.keys(queries).reduce((acc, withinQueryName) => {
  return {
    ...acc,
    [withinQueryName]: new Function(`
    const els = arguments[0];
    if(els.length > 1) {
      throw new Error("within() only works with a single element, found " + els.length);
    }
    const el = els[0];
    const args = Array.from(arguments).slice(1);
    return window.TestingLibraryDom.within(el).${withinQueryName}.apply(null, args);
  `),
  };
}, {} as Record<keyof typeof queries, (node: Element, ...methodParams: any[]) => any>);

export async function configureOnce(options: Partial<Options>) {
  const { content } = configure(options);
  // @ts-ignore
  await ClientFunction(new Function(content))();
}

export function configure(options: Partial<Options>) {
  const configFunction = `
  window.TestingLibraryDom.configure(${JSON.stringify(options)});
`;
  return { content: configFunction };
}

export function within<T>(
  sel: string | Selector | SelectorPromise | TestcafeBoundFunction<T>
): TestcafeBoundFunctions<typeof queries> {
  if (sel instanceof Function) {
    return within(sel());
  }
  if (isSelector(sel)) {
    // @ts-ignore
    return sel.addCustomMethods(withinSelectors, { returnDOMNodes: true });
  } else if (typeof sel === "string") {
    return within(Selector(sel));
  } else {
    throw new Error(
      `"within" only accepts a query (getBy, queryBy, etc), string or testcafe Selector`
    );
  }
}

function isSelector(sel: any): sel is Selector {
  return sel.constructor.name === SELECTOR_TYPE;
}

const bindFunction = <T extends keyof typeof queries>(queryName: T) => {
  const query = queryName.replace("find", "query") as T;
  return Selector(
    (matcher, ...options) => {
      return window.TestingLibraryDom[query](
        document.body,
        matcher,
        ...options
      ) as Node | Node[] | NodeList | HTMLCollection;
    },
    {
      dependencies: { query },
    }
  );
};

export const getByLabelText = bindFunction("getByLabelText");
export const getAllByLabelText = bindFunction("getAllByLabelText");
export const queryByLabelText = bindFunction("queryByLabelText");
export const queryAllByLabelText = bindFunction("queryAllByLabelText");
export const findByLabelText = bindFunction("findByLabelText");
export const findAllByLabelText = bindFunction("findAllByLabelText");
export const getByPlaceholderText = bindFunction("getByPlaceholderText");
export const getAllByPlaceholderText = bindFunction("getAllByPlaceholderText");
export const queryByPlaceholderText = bindFunction("queryByPlaceholderText");
export const queryAllByPlaceholderText = bindFunction(
  "queryAllByPlaceholderText"
);
export const findByPlaceholderText = bindFunction("findByPlaceholderText");
export const findAllByPlaceholderText = bindFunction(
  "findAllByPlaceholderText"
);
export const getByText = bindFunction("getByText");
export const getAllByText = bindFunction("getAllByText");
export const queryByText = bindFunction("queryByText");
export const queryAllByText = bindFunction("queryAllByText");
export const findByText = bindFunction("findByText");
export const findAllByText = bindFunction("findAllByText");
export const getByAltText = bindFunction("getByAltText");
export const getAllByAltText = bindFunction("getAllByAltText");
export const queryByAltText = bindFunction("queryByAltText");
export const queryAllByAltText = bindFunction("queryAllByAltText");
export const findByAltText = bindFunction("findByAltText");
export const findAllByAltText = bindFunction("findAllByAltText");
export const getByTitle = bindFunction("getByTitle");
export const getAllByTitle = bindFunction("getAllByTitle");
export const queryByTitle = bindFunction("queryByTitle");
export const queryAllByTitle = bindFunction("queryAllByTitle");
export const findByTitle = bindFunction("findByTitle");
export const findAllByTitle = bindFunction("findAllByTitle");
export const getByDisplayValue = bindFunction("getByDisplayValue");
export const getAllByDisplayValue = bindFunction("getAllByDisplayValue");
export const queryByDisplayValue = bindFunction("queryByDisplayValue");
export const queryAllByDisplayValue = bindFunction("queryAllByDisplayValue");
export const findByDisplayValue = bindFunction("findByDisplayValue");
export const getByRole = bindFunction("getByRole");
export const getAllByRole = bindFunction("getAllByRole");
export const queryByRole = bindFunction("queryByRole");
export const queryAllByRole = bindFunction("queryAllByRole");
export const findByRole = bindFunction("findByRole");
export const findAllByRole = bindFunction("findAllByRole");
export const findAllByDisplayValue = bindFunction("findAllByDisplayValue");
export const getByTestId = bindFunction("getByTestId");
export const getAllByTestId = bindFunction("getAllByTestId");
export const queryByTestId = bindFunction("queryByTestId");
export const queryAllByTestId = bindFunction("queryAllByTestId");
export const findByTestId = bindFunction("findByTestId");
export const findAllByTestId = bindFunction("findAllByTestId");

export * from "./types";
