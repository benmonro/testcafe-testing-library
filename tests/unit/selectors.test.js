import * as selectors from '../../src'

/* eslint-enable import/namespace */

test('exports expected selectors', () => {
  expect(selectors).toMatchObject(expect.any(Object))
  expect(Object.keys(selectors)).toMatchSnapshot()
  Object.keys(selectors).forEach(selector => {
    // eslint-disable-next-line import/namespace
    expect(selectors[selector]).toBeInstanceOf(Function)
  })
})
