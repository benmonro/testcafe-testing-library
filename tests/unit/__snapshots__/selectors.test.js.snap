// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exports expected selectors 1`] = `
Array [
  "configure",
  "addTestcafeTestingLibrary",
  "within",
  "queryAllByLabelText",
  "getAllByLabelText",
  "queryByLabelText",
  "getByLabelText",
  "findAllByLabelText",
  "findByLabelText",
  "queryByPlaceholderText",
  "queryAllByPlaceholderText",
  "getByPlaceholderText",
  "getAllByPlaceholderText",
  "findAllByPlaceholderText",
  "findByPlaceholderText",
  "queryAllByText",
  "queryByText",
  "getByText",
  "getAllByText",
  "findAllByText",
  "findByText",
  "queryAllByDisplayValue",
  "queryByDisplayValue",
  "getByDisplayValue",
  "getAllByDisplayValue",
  "findAllByDisplayValue",
  "findByDisplayValue",
  "queryAllByAltText",
  "queryByAltText",
  "getByAltText",
  "getAllByAltText",
  "findAllByAltText",
  "findByAltText",
  "queryAllByTitle",
  "queryByTitle",
  "getByTitle",
  "getAllByTitle",
  "findAllByTitle",
  "findByTitle",
  "queryByRole",
  "queryAllByRole",
  "getAllByRole",
  "getByRole",
  "findAllByRole",
  "findByRole",
  "queryByTestId",
  "queryAllByTestId",
  "getByTestId",
  "getAllByTestId",
  "findAllByTestId",
  "findByTestId",
]
`;
